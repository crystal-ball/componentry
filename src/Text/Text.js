import elem from '../elem-factory'
import { useTheme } from '../Theme/Theme'

/**
 * Text component
 * TODO: use `span` prop to render a span?
 */
export default function Text(props) {
  const merged = { as: 'p', ...useTheme('Text'), ...props }
  return elem({ componentClassNames: merged.as, ...merged })
}
