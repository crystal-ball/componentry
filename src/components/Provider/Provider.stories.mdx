import { Meta } from '@storybook/addon-docs'

import { Button } from '../Button/Button'
import { ComponentryProvider } from './Provider'
import { ProviderConsumer } from './storybook/ProviderConsumer'

<Meta title='Components/ComponentryProvider' component={ComponentryProvider} />

# Theme

<ComponentryProvider
  components={{
    Button: {
      color: 'success',
      size: 'lg',
    },
  }}
>
  <div>
    <Button>A large successful button!</Button>
    <ProviderConsumer />
  </div>
</ComponentryProvider>
