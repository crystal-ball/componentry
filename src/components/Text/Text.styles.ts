import { type CSSProperties } from 'react'
import { Theme } from '../../theme/theme'

//                                         <Text /> styles
// -------------------------------------------------------

export const textStyles = (theme: Theme): TextStyles => ({
  // BASE
  '.C9Y-Text-base': {},

  // VARIANTS
  '.C9Y-Text-h1': {
    fontSize: theme.fontSize.h1,
    color: theme.colors.gray[900],
  },
  '.C9Y-Text-h2': {
    fontSize: theme.fontSize.h2,
    color: theme.colors.gray[900],
  },
  '.C9Y-Text-h3': {
    fontSize: theme.fontSize.h3,
    color: theme.colors.gray[900],
  },
  '.C9Y-Text-body': {
    fontSize: theme.fontSize.body,
    color: theme.colors.gray[800],

    // Set spacing between multiple paragraphs using sibling selector and
    // margin-top.
    '& + &': {
      marginTop: theme.spacing[4],
    },
  },
})

export interface TextStyles {
  /** Base class applied to all variants for shared structural styles */
  '.C9Y-Text-base': CSSProperties
  /** Variant class applied when `variant="h1"` */
  '.C9Y-Text-h1': CSSProperties
  /** Variant class applied when `variant="h2"` */
  '.C9Y-Text-h2': CSSProperties
  /** Variant class applied when `variant="h3"` */
  '.C9Y-Text-h3': CSSProperties
  /** Variant class applied when `variant="body"` */
  '.C9Y-Text-body': {
    /** Sibling selector for auto-spacing multiple body elements */
    '& + &': CSSProperties
  } & CSSProperties
}
