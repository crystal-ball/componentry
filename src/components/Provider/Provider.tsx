/* eslint-disable no-underscore-dangle */
import { type ReactElement, createContext, useContext, useMemo } from 'react'

import { type Theme } from '../../theme/theme'
import { themeDefaults } from '../../theme/theme-defaults'
import { initializeUtilityClassesTheme } from '../../utils/utility-classes'

import {
  type ActiveActionProps,
  type ActiveContentProps,
  type ActiveProps,
} from '../Active/Active'
import { type AlertProps } from '../Alert/Alert'
import { type BadgeProps } from '../Badge/Badge'
import { type BlockProps } from '../Block/Block'
import { type ButtonProps } from '../Button/Button'
import { type CardProps } from '../Card/Card'
import {
  type DrawerActionProps,
  type DrawerContentProps,
  type DrawerProps,
} from '../Drawer/Drawer'
import {
  type DropdownActionProps,
  type DropdownContentProps,
  type DropdownItemProps,
  type DropdownProps,
} from '../Dropdown/Dropdown'
import { type FlexProps } from '../Flex/Flex'
import { type GridProps } from '../Grid/Grid'
import { type IconProps } from '../Icon/Icon'
import { type IconButtonProps } from '../IconButton/IconButton'
import { type InputFieldProps, type InputLabelProps } from '../Input/Input'
import { type LinkProps } from '../Link/Link'
import {
  type ModalBodyProps,
  type ModalHeaderProps,
  type ModalProps,
  type ModalTitleProps,
} from '../Modal/Modal'
import {
  type PopoverActionProps,
  type PopoverContentProps,
  type PopoverProps,
} from '../Popover/Popover'
import {
  type TabsActionProps,
  type TabsActionsContainerProps,
  type TabsContentProps,
  type TabsProps,
} from '../Tabs/Tabs'
import { type TextProps } from '../Text/Text'
import {
  type TooltipActionProps,
  type TooltipContentProps,
  type TooltipProps,
} from '../Tooltip/Tooltip'

export type Components = {
  Active?: ActiveProps
  ActiveAction?: ActiveActionProps
  ActiveContent?: ActiveContentProps
  Alert?: AlertProps
  Badge?: BadgeProps
  Block?: BlockProps
  Button?: ButtonProps
  Card?: CardProps
  Drawer?: DrawerProps
  DrawerAction?: DrawerActionProps
  DrawerContent?: DrawerContentProps
  Dropdown?: DropdownProps
  DropdownItem?: DropdownItemProps
  DropdownAction?: DropdownActionProps
  DropdownContent?: DropdownContentProps
  Flex?: FlexProps
  Grid?: GridProps
  Icon?: IconProps
  IconButton?: IconButtonProps
  InputField?: InputFieldProps
  InputLabel?: InputLabelProps
  Link?: LinkProps
  Modal?: ModalProps
  ModalBody?: ModalBodyProps
  ModalHeader?: ModalHeaderProps
  ModalTitle?: ModalTitleProps
  Popover?: PopoverProps
  PopoverAction?: PopoverActionProps
  PopoverContent?: PopoverContentProps
  Tabs?: TabsProps
  TabsAction?: TabsActionProps
  TabsContent?: TabsContentProps
  TabsActionsContainer?: TabsActionsContainerProps
  Text?: TextProps
  Tooltip?: TooltipProps
  TooltipAction?: TooltipActionProps
  TooltipContent?: TooltipContentProps
}

/** Componentry Context */
const ComponentryCtx = createContext<null | { components: Components; theme: Theme }>(
  null,
)

interface ComponentryProviderProps {
  children: ReactElement
  /** Component default props */
  components?: Components
  /** Application theme values */
  theme?: Theme
}

export interface ComponentryProvider {
  (props: ComponentryProviderProps): JSX.Element
  displayName?: string
}

/**
 * #### [üìù ComponentryProvider](https://componentry.design/components/provider)
 *
 * Provider for the application theme and default component props.
 * @example
 * const appTheme = {}
 * const defaultProps = {}
 * ```tsx
 * <ComponentryProvider theme={appTheme} components={defaultProps}>
 *  <App />
 * </ComponentryProvider>
 * ```
 */
export const ComponentryProvider: ComponentryProvider = ({
  children,
  components,
  theme,
}) => {
  const contextValue = useMemo(() => {
    return { components: components ?? {}, theme: theme ?? themeDefaults }
  }, [components, theme])

  // Internal convenience helper: if user has provided a theme value initialize
  // the utility classes module with it for them
  initializeUtilityClassesTheme(contextValue.theme)

  return (
    <ComponentryCtx.Provider value={contextValue}>{children}</ComponentryCtx.Provider>
  )
}
ComponentryProvider.displayName = 'ComponentryProvider'

// --------------------------------------------------------
// THEME

/**
 * [Theme hook üìù](https://componentry.design/components/theme)
 */
export function useTheme(): Theme {
  const ctx = useContext(ComponentryCtx)
  return ctx ? ctx.theme : themeDefaults
}

// --------------------------------------------------------
// PROPS

let preCompileMode = false
let preCompileComponentsValue: Components = {}

export function __initializePreCompileMode(components: Components): void {
  preCompileMode = true
  preCompileComponentsValue = components
}

function useContextProps<Name extends keyof Components>(
  componentName: Name,
): Components[Name] {
  const ctx = useContext(ComponentryCtx)
  return ctx ? ctx.components[componentName] : undefined
}

/**
 * Internal function for accessing component default props through context.
 */
export function useThemeProps<Name extends keyof Components>(
  componentName: Name,
): Components[Name] {
  if (preCompileMode) {
    return preCompileComponentsValue[componentName]
  }

  // During Babel pre-compiling `preCompileMode` will always be true, during
  // runtime execution it will always be false
  // eslint-disable-next-line react-hooks/rules-of-hooks
  return useContextProps(componentName)
}
