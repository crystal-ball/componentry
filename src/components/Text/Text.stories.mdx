import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs'

import { Block } from '../Block/Block'
import { Text } from './Text'

<Meta title='Components/Text' component={Text} />

# Text

## Variants

Componentry ships with support for variants: `h{1,3}`, `body`, and `small`.

<Story name='Text'>
  <Text variant='h1'>Heading 1</Text>
  <Text variant='h2'>Heading 2</Text>
  <Text variant='h3'>Heading 3</Text>
  <Text variant='body'>
    Life finds a way. Life finds a way. God help us, we&apos;re in the hands of engineers.
    Life finds a way. Must go faster. Hey, you know how I&apos;m, like, always trying to
    save the planet? Here&apos;s my chance. Eventually, you do plan to have dinosaurs on
    your dinosaur tour, right?
  </Text>
  <Text variant='small'>Small copy</Text>
</Story>

## Body spacing

By default Componentry includes margin-top between sibling body elements, this can be
configured using the `& + &` stle in the body variant styles.

<Text variant='body'>
  Life finds a way. Life finds a way. God help us, we&apos;re in the hands of engineers.
  Life finds a way. Must go faster. Hey, you know how I&apos;m, like, always trying to
  save the planet? Here&apos;s my chance. Eventually, you do plan to have dinosaurs on
  your dinosaur tour, right?
</Text>
<Text variant='body'>
  Life finds a way. Life finds a way. God help us, we&apos;re in the hands of engineers.
  Life finds a way. Must go faster. Hey, you know how I&apos;m, like, always trying to
  save the planet? Here&apos;s my chance. Eventually, you do plan to have dinosaurs on
  your dinosaur tour, right?
</Text>

## Truncating

Truncate overflowing text with `truncate`

<Block width={240}>
  <Text truncate>
    Eventually, you do plan to have dinosaurs on your dinosaur tour, right?
  </Text>
</Block>

## Utility styles - fontSize

The `fontSize` prop is intended to be used with the `body` variant:

<Text variant='body' fontSize='sm'>
  Small size body.
</Text>
<Text variant='body' fontSize='base'>
  Base size body.
</Text>
<Text variant='body' fontSize='lg'>
  Large size body.
</Text>

## Utility styles - bold & italic

Shorthand `bold` and `italic` props are available:

<Text italic>Italic text</Text>
<Text bold>Bold text</Text>

## Props

<ArgsTable of={Text} />
