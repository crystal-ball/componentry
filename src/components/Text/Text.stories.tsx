import type { Meta, StoryObj } from '@storybook/react'

import { Block } from '../Block/Block'
import { Text } from './Text'

const meta: Meta<typeof Text> = { component: Text }

export default meta
type Story = StoryObj<typeof Text>

export const Primary: Story = {
  render: () => (
    <>
      <Text variant='h1'>Heading 1</Text>

      <Text variant='h2'>Heading 2</Text>
      <Text variant='h3'>Heading 3</Text>
      <Text variant='body'>
        Life finds a way. Life finds a way. God help us, we&apos;re in the hands of
        engineers. Life finds a way. Must go faster. Hey, you know how I&apos;m, like,
        always trying to save the planet? Here&apos;s my chance. Eventually, you do plan
        to have dinosaurs on your dinosaur tour, right?
      </Text>
      <Text variant='small'>Small copy</Text>
    </>
  ),
}

export const BodySpacing: Story = {
  render: () => (
    <>
      <Text variant='body'>
        By default Componentry includes margin-top between sibling body elements, this can
        be configured using the `& + &` stle in the body variant styles.
      </Text>
      <Text variant='body'>
        By default Componentry includes margin-top between sibling body elements, this can
        be configured using the `& + &` stle in the body variant styles.
      </Text>
    </>
  ),
}

export const Truncation: Story = {
  render: () => (
    // @ts-expect-error -- storybook types
    <Block width={240}>
      <Text truncate>
        Eventually, you do plan to have dinosaurs on your dinosaur tour, right?
      </Text>
    </Block>
  ),
}

export const FontSize: Story = {
  render: () => (
    <>
      <Text variant='body' fontSize='sm'>
        Small size body.
      </Text>
      <Text variant='body' fontSize='base'>
        Base size body.
      </Text>
      <Text variant='body' fontSize='lg'>
        Large size body.
      </Text>
    </>
  ),
}

export const BoldItalic: Story = {
  render: () => (
    <>
      <Text italic>Italic text</Text>
      <Text bold>Bold text</Text>
    </>
  ),
}
