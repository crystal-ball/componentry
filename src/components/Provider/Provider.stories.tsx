import type { Meta, StoryObj } from '@storybook/react'

import { Button } from '../Button/Button'
import { ComponentryProvider, useTheme } from './Provider'

const meta: Meta<typeof ComponentryProvider> = {
  component: ComponentryProvider,
}

export default meta
type Story = StoryObj<typeof ComponentryProvider>

export const Primary: Story = {
  render: () => (
    <ComponentryProvider
      config={{
        defaultProps: {
          Button: {
            size: 'large',
          },
        },
      }}
    >
      <div>
        <Button>A large successful button!</Button>
        <ProviderConsumer />
      </div>
    </ComponentryProvider>
  ),
}

function ProviderConsumer() {
  const theme = useTheme()

  return (
    <div>
      <h3>Theme values:</h3>
      <pre style={{ width: 600 }}>
        <code>{JSON.stringify(theme, null, 2)}</code>
      </pre>
    </div>
  )
}
