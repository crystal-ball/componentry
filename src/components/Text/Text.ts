import React, { forwardRef } from 'react'
import { element } from '../../utils/element-creator'
import { MergeTypes, Resolve } from '../../utils/types'
import { UtilityProps } from '../../utils/utility-classes'
import { useThemeProps } from '../Provider/Provider'

// --------------------------------------------------------
// TEXT ELEMENTS MAP

/**
 * Mapping of Text variants to rendered elements
 * @example
 * ```ts
 * const textElementsMap: TextElementsMap = {
 *   h1: 'h1',
 *   body: 'p',
 * }
 * ```
 */
export type TextElementsMap = {
  [Variant: string]: keyof JSX.IntrinsicElements | React.ComponentType<unknown>
}
/**
 * Internal map used for final rendering
 * @see {@link configureTextElementsMap}
 */
let textElementMap: TextElementsMap = {
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
  body: 'div',
  code: 'code',
  small: 'small',
}

/**
 * Configuration method for defining the elements to render for each text
 * variant.
 * @remarks
 * Configured elements are merged with the library default values.
 * @example
 * ```ts
 * // Add a title variant that renders an h1, and render a div instead of a p
 * // for body variants.
 * configureTextElementsMap({
 *   title: 'h1',
 *   body: 'div',
 * })
 * ```
 */
export function configureTextElementsMap(elementsMap: TextElementsMap): void {
  textElementMap = { ...textElementMap, ...elementsMap }
}

// --------------------------------------------------------
// TEXT COMPONENT

/** Module augmentation interface for overriding component props' types */
export interface TextPropsOverrides {}

export interface TextPropsDefaults {
  variant?: 'h1' | 'h2' | 'h3' | 'body' | 'code' | 'small'
  /** Truncates overflowing text with an ellipses */
  truncate?: boolean
}

export type TextProps<Elem extends React.ElementType = 'div'> = Resolve<
  MergeTypes<TextPropsDefaults, TextPropsOverrides>
> &
  UtilityProps &
  React.ComponentPropsWithRef<Elem> & { as?: Elem }

// ‚ú® Nice display type for IntelliSense
export interface Text {
  <Elem extends React.ElementType = 'div'>(
    props: TextProps<Elem>,
  ): React.ReactElement | null
  displayName?: string
}

/**
 * Text provides consistently themed typography elements.
 * @example
 * ```tsx
 * <Text variant="h1">
 *   Build something delightful!
 * </Text>
 * ```
 * @see [üìù Text docs](https://componentry.design/docs/components/text)
 */
export const Text: Text = forwardRef<HTMLDivElement, TextProps>((props, ref) => {
  const {
    truncate = false,
    variant = 'body',
    ...rest
  } = {
    ...useThemeProps('Text'),
    ...props,
  }

  return element({
    ref,
    as: textElementMap[variant],
    componentCx: [`C9Y-Text-base C9Y-Text-${variant}`, { truncate }],
    ...rest,
  })
})
Text.displayName = 'Text'
