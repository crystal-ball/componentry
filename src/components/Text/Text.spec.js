import { render, screen } from '@testing-library/react'

import { elementTests } from '../../test/element-tests'
import { ComponentryProvider } from '../Provider/Provider'
import { Text } from './Text'

describe('<Text/>', () => {
  // Basic library element test suite
  elementTests(Text)
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Text /> snapshots', () => {
  it('renders correctly', () => {
    render(<Text>Componentry</Text>)

    expect(screen.getByText('Componentry')).toMatchSnapshot()
  })
})

// --------------------------------------------------------
// Configuration

describe('Text', () => {
  it('ComponentryProvider allows configuring variant render elements', () => {
    render(
      <ComponentryProvider
        config={{
          defaultProps: {
            Text: {
              textElementMap: {
                rad: 'section',
              },
            },
          },
        }}
      >
        <Text variant='rad'>Componentry</Text>
      </ComponentryProvider>,
    )

    expect(screen.getByText('Componentry')).toContainHTML(
      '<section class="C9Y-Text-base C9Y-Text-rad">Componentry</section>',
    )
  })
})
