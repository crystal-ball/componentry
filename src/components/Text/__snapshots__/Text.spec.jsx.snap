// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Text /> snapshots > renders correctly 1`] = `
<div
  class="C9Y-Text-base C9Y-Text-body"
>
  Componentry
</div>
`;
